rascal-lsp/src/main/rascal/lang/rascal/tests/rename/TestUtils.rsc

```diff
@@ -0,0 +1,288 @@
+    occs = extractRenameOccurrences(singleModule, edits, oldName);
+
+    for (src <- pcfg.srcs) {
+        verifyTypeCorrectRenaming(src, edits, pcfg);
+    }
+
+    return <edits, occs>;
+}
+
+list[DocumentEdit] getEdits(str stmtsStr, int cursorAtOldNameOccurrence, str oldName, str newName, str decls, str imports) {
+    <edits, _, _> = getEditsAndModule(stmtsStr, cursorAtOldNameOccurrence, oldName, newName, decls, imports);
+    return edits;
+}
+
+private tuple[list[DocumentEdit], set[int], loc] getEditsAndModule(str stmtsStr, int cursorAtOldNameOccurrence, str oldName, str newName, str decls, str imports, str moduleName = "TestModule<abs(arbInt())>") {
+    str moduleStr =
+    "module <moduleName>
+    '<trim(imports)>
+    '<trim(decls)>
+    'void main() {
+    '<trim(stmtsStr)>
+    '}";
+
+    // Write the file to disk (and clean up later) to easily emulate typical editor behaviour
+    loc testDir = |memory://tests/rename/<moduleName>|;
+    loc moduleFileName = testDir + "rascal" + "<moduleName>.rsc";
+    writeFile(moduleFileName, moduleStr);
+
+    <edits, occs> = getEditsAndOccurrences(moduleFileName, testDir, cursorAtOldNameOccurrence, oldName, newName);
+    return <edits, occs, moduleFileName>;
+}
+
+private set[int] extractRenameOccurrences(loc moduleFileName, list[DocumentEdit] edits, str name) {
+    start[Module] m = parseModuleWithSpaces(moduleFileName);
+    list[loc] oldNameOccurrences = [];
+    for (/Name n := m, "<n>" == name) {
+        oldNameOccurrences += n.src;
+    }
+
+    if ([changed(_, replaces)] := edits) {
+        idx = {indexOf(oldNameOccurrences, l) | replace(l, _) <- replaces};
+        return idx;
+    } else {
+        print("Unexpected changes: ");
+        iprintln(edits);
+        throw "Unexpected changes: <edits>";
+    }
+}
+
+private str moduleNameToPath(str name) = replaceAll(name, "::", "/");
+private str modulePathToName(str path) = replaceAll(path, "/", "::");
+
+private Tree findCursor(loc f, str id, int occ) {
+    m = parseModuleWithSpaces(f);
+    return [n | /Name n := m.top, "<n>" == id][occ];
+}
+
+private loc storeTestModule(loc dir, str name, str body) {
+    str moduleStr = "
+    'module <name>
+    '<body>
+    ";
+
+    loc moduleFile = dir + "rascal" + (moduleNameToPath(name) + ".rsc");
+    writeFile(moduleFile, moduleStr);
+
+    return moduleFile;
+}
+
+private set[Tree] occsToTrees(start[Module] m, str name, set[int] occs) = {n | i <- occs, n := [n | /Name n := m.top, "<n>" == name][i]};
+private set[loc] occsToLocs(start[Module] m, str name, set[int] occs) = {t.src | t <- occsToTrees(m, name, occs)};
+private set[int] locsToOccs(start[Module] m, str name, set[loc] occs) = {indexOf(names, occ) | names := [n.src | /Name n := m.top, "<n>" == name], occ <- occs};
+
+// Workaround to be able to pattern match on the emulated `src` field
+data Tree (loc src = |unknown:///|(0,0,<0,0>,<0,0>));
```

Comment:
This workaround appears multiple times. Consider moving it to a single location

Reply:
Removed from the test utils, as that imports it from the other location